thunx-encoding-json/src/main/java/com/contentgrid/thunx/encoding/json/JsonThunkExpressionCoder.java

```diff
@@ -102,6 +98,84 @@ public JsonExpressionDto visit(Variable variable) {
+        protected JsonExpressionDto visit(SetValue setValue) {
+            return processSetValueDto(setValue);
+        }
+
+        @Override
+        protected JsonExpressionDto visit(ListValue listValue) {
+            return processListValueDto(listValue);
+        }
+
+
+    }
+
+    private static class JsonScalarEncoderVisitor extends ContextFreeThunkExpressionVisitor<JsonScalarDto<?>> {
+
+        @Override
+        protected JsonScalarDto<?> visit(Scalar<?> scalar) {
+            return processScalar(scalar);
+        }
+
+        @Override
+        protected JsonScalarDto<?> visit(FunctionExpression<?> functionExpression) {
+            throw new UnsupportedOperationException("Only Scalars are supported in Collection.");
+        }
+
+        @Override
+        protected JsonScalarDto<?> visit(SymbolicReference symbolicReference) {
+            throw new UnsupportedOperationException("Only Scalars are supported in Collection.");
+        }
+
+        @Override
+        protected JsonScalarDto<?> visit(Variable variable) {
+            throw new UnsupportedOperationException("Only Scalars are supported in Collection.");
+        }
+
+        @Override
+        protected JsonScalarDto<?> visit(SetValue setValue) {
+            throw new UnsupportedOperationException("Only Scalars are supported in Collection.");
+        }
+
+        @Override
+        protected JsonScalarDto<?> visit(ListValue listValue) {
+            throw new UnsupportedOperationException("Only Scalars are supported in Collection.");
+        }
+
+    }
+
+    private static JsonScalarDto<?> processScalar(Scalar<?> scalar) {
+        var resultType = scalar.getResultType();
+
+        var typeName = JsonScalarDto.SCALAR_TYPES.get(resultType);
+        if (typeName == null) {
+            String exMessage = String.format("Scalar of type <%s> is not supported", resultType.getName());
+            throw new IllegalArgumentException(exMessage);
+        }
+
+        return JsonScalarDto.of(typeName, scalar.getValue());
+    }
+
+    private static JsonExpressionDto processSetValueDto(SetValue setValue) {
+        var resultType = setValue.getResultType();
+
+        Collection<JsonExpressionDto> result = setValue.getValue()
+                    .stream()
+                    .map(expression -> expression.accept(new JsonEncoderVisitor(), null))
+                    .collect(Collectors.toSet());
+
+        return JsonCollectionValueDto.of(JsonCollectionValueDto.getTypeClass(resultType), result);
+    }
+
+    private static JsonExpressionDto processListValueDto(ListValue listValue) {
+        var resultType = listValue.getResultType();
+
+        Collection<JsonExpressionDto> result = listValue.getValue()
+                    .stream()
+                    .map(scalar -> scalar.accept(new JsonEncoderVisitor(), null))
+                    .collect(Collectors.toList());
 
+        return JsonCollectionValueDto.of(JsonCollectionValueDto.getTypeClass(resultType), result);
```

Comment:
These methods each have only one usage in the visitor above.

You can place the implementations there instead of having a separate function call